### PR TITLE
Fix for off-by-one error reported in #64.

### PR DESCRIPTION
**Pull request by demiankatz**
- Includes new offline test for Stream class.

### CODE REVIEW

File: tests/ZendService/Amazon/S3/OfflineStreamTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            )->will($this->returnValue($response2));
+
+        // Register the mock client so it is called for all test:// URLs:
+        $s3->registerAsClient('test');
+
+        // Run the test:
+        $stream = new S3\Stream();
+        $stream->stream_open('test://foo/bar', 'r', 0, null);
+        $stream->stream_read(5000);
+        $stream->stream_read(5000);
```
Comment (Xerkus) [inline tests/ZendService/Amazon/S3/OfflineStreamTest.php]: count that returned data is exactly 8000 bytes

**Comment by Xerkus** [comment]
There is number of problems with the test.
Range header test is pushed to setup function, document is too small to detect breakage: should you remove Range assert, asserting equality of result and the original will not fail.
Interesting, if i remove header assert and increase document size, `testFread()` won't fail but `testStreamRead()` will. Test mock is returning proper ranges, not the actually requested

**Comment by demiankatz** [comment]
@Xerkus,

I have revised the test so that it has fewer hard-coded assumptions. You can now change the size of the document in setUpS3() and the tests will continue to perform their jobs. I've made the test document a bit larger by default. I didn't use a memory stream because it didn't seem to be necessary, but I followed the suggested principles you suggested on #64. As I noted there, the disadvantage to this more flexible approach is that when the test fails (and I did confirm that this test fails in master), you get a long and puzzling string mismatch error instead of an indication that a bad header was generated. I can live with that if you can.

**Comment by Xerkus** [comment]
I am working on a setup method right now, going to push it in a bit to my fork  with a full example of what i meant

**Comment by demiankatz** [comment]
@Xerkus, I'm heading home now but can discuss this further tomorrow. Please compare the work I have done against whatever you are working on -- I've just bumped up the size of the test document a bit more to ensure that it is large enough to demonstrate a problem with the default fread chunk size, in addition to the artificially small 500-byte chunk size in the more "manual" test.

**Comment by Xerkus** [comment]
Updated test at https://github.com/Xerkus/ZendService_Amazon/blob/hotfix/off-by-one-bug/tests/ZendService/Amazon/S3/OfflineStreamTest.php
Conclusively shows the bug, before fix:
```
PHPUnit 4.8.35 by Sebastian Bergmann and contributors.

...............................................................  63 / 184 ( 34%)
...............................................SSSSSSSSSSSSSSFF 126 / 184 ( 68%)
.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.........SSSSSSSSSSSSSS

Time: 2.35 seconds, Memory: 10.00MB

There were 2 failures:

1) ZendServiceTest\Amazon\S3\OfflineStreamTest::testFread
Resulting data does not match original
Failed asserting that false is true.

tests/ZendService/Amazon/S3/OfflineStreamTest.php:120

2) ZendServiceTest\Amazon\S3\OfflineStreamTest::testStreamRead
Resulting data does not match original
Failed asserting that false is true.

tests/ZendService/Amazon/S3/OfflineStreamTest.php:148
```

Test is missing another test case for the proper `Range` header values

**Comment by demiankatz** [comment]
@Xerkus, thanks for taking the time to make these significant improvements to the test! I have gone ahead and created another test that checks that the range values are set correctly on the first chunk requested. Obviously we could take this further and test range values on subsequent chunks, since there could be different error conditions there -- but I'm not sure if it's worth investing more time in at this point. I'd be content to merge this now, unless you really feel that further test improvements are needed.

**Comment by Xerkus** [comment]
@demiankatz not necessary, it can be a callback constraint.

Everything is good for the scope of this issue

**Comment by Xerkus** [comment]
@demiankatz thank you

**Pull request merged**